plugins/channel_hint.c

```diff
@@ -0,0 +1,97 @@
+#include "config.h"
+#include <plugins/channel_hint.h>
+
+void json_add_channel_hint(struct json_stream *js,
+			   const char *fieldname,
+			   const struct channel_hint *hint)
+{
+	json_object_start(js, fieldname);
+	json_add_u32(js, "timestamp", hint->timestamp);
+	json_add_short_channel_id_dir(js, "sciddir", &hint->scidd);
+	json_add_amount_msat(js, "estimated_capacity_msat",
+			     hint->estimated_capacity);
+	json_add_amount_msat(js, "overall_capacity_msat",
+			     hint->overall_capacity);
+	json_add_bool(js, "enabled", hint->enabled);
+	json_object_end(js);
+}
+
+#define PAY_REFILL_TIME 7200
+
+/**
+ * Update the `channel_hint` in place, return whether it should be kept.
+ *
+ * This computes the refill-rate based on the overall capacity, and
+ * the time elapsed since the last update and relaxes the upper bound
+ * on the capacity, and resets the enabled flag if appropriate. If the
+ * hint is no longer useful, i.e., it does not provide any additional
+ * information on top of the structural information we've learned from
+ * the gossip, then we return `false` to signal that the
+ * `channel_hint` may be removed.
+ */
+bool channel_hint_update(const struct timeabs now, struct channel_hint *hint)
+{
+	/* Precision is not required here, so integer division is good
+	 * enough. But keep the order such that we do not round down
+	 * too much. We do so by first multiplying, before
+	 * dividing. The formula is `current = last + delta_t *
+	 * overall / refill_rate`.
+	 */
+	struct amount_msat refill;
+	u64 seconds = now.ts.tv_sec - hint->timestamp;
+	if (!amount_msat_mul(&refill, hint->overall_capacity, seconds))
+		abort();
+
+	refill = amount_msat_div(refill, PAY_REFILL_TIME);
+	if (!amount_msat_add(&hint->estimated_capacity,
+			     hint->estimated_capacity, refill))
+		abort();
+
+	/* Clamp the value to the `overall_capacity` */
+	if (amount_msat_greater(hint->estimated_capacity,
+				hint->overall_capacity))
+		hint->estimated_capacity = hint->overall_capacity;
```

Comment:
I think it is unlikely to hit an overflow under normal conditions. For example a timeframe of 1 year ~3e7 seconds times a channel capacity of 1 BTC = 1e11 msat, when multiplied gives ~3e18 that fits into 64 bits.

I would anyways do the computation only if `seconds` is below `PAY_REFILL_TIME` 'cause otherwise we know immediately that the estimated capacity has been refilled completely:
```suggestion
	if (seconds >= PAY_REFILL_TIME) {
		/* no computation is needed, no overflow */
		hint->estimated_capacity = hint->overall_capacity;
	} else {
		if (!amount_msat_mul(&refill, hint->overall_capacity, seconds))
			abort();

		refill = amount_msat_div(refill, PAY_REFILL_TIME);
		if (!amount_msat_add(&hint->estimated_capacity,
				     hint->estimated_capacity, refill))
			abort();

		/* Clamp the value to the `overall_capacity` */
		hint->estimated_capacity = amount_msat_min(
		    hint->overall_capacity, hint->estimated_capacity);
	}
```

Reply:
I don't think there is a logical error in the current code? Are you mostly looking at optimizations? In that case I'd defer to a separate PR, since it is not wrong per se, maybe suboptimal.